Add GEOS-Chem species into Fast-JX

xk-y opened the pull request.

xk-y commented:
I have reviewed and hereby sign the CLA

jialinl6 commented:
> We can confirm that the units in the GEOS-Chem full chemistry is using molecules/cc [here](https://github.com/geoschem/geos-chem/blob/f4c75742e13da440b34c4968797aa150ce114113/GeosCore/fullchem_mod.F90#L376) and [here](https://github.com/geoschem/geos-chem/blob/f4c75742e13da440b34c4968797aa150ce114113/GeosCore/fullchem_mod.F90#L1658), which are two calls of unit conversion in the Do_FullChem subroutine. Each time it calls chemistry mechanism, the species units shall be converted into molecules/cc for the purpose of KPP, and then converted back to original units after finishing chemistry.
> 
> I checked our [geoschem_ratelaws.jl](https://github.com/EarthSciML/GasChem.jl/blob/73ca1aec53bc3ce326801b94d9af15b85a55d2bb/src/geoschem_ratelaws.jl) and the reaction rate constants are the same as the ones in the GEOS-Chem, meaning that we should set all species units as molecules/cc rather than ppb at least in our chemistry module. A potential solution is to create a unit conversion module acting as [the ones in GEOS-Chem](https://github.com/geoschem/geos-chem/blob/main/GeosUtil/unitconv_mod.F90), to ensure we are in molecules/cc in chemistry, and ppb in other modules such as advection.
> 
> I also checked this pull request. There are only first-order reactions for photolysis, with rate constant unit as s^-1, meaning that the unit of species does not affect the chemistry, and the ratio of photolyzed species is the same. That's why the photolysis looks reasonable.

ctessum commented:
Ok, it sounds like the units issue doesn't affect the current pull request, but it's still an issue we need to resolve. What are the units that are used outside of kpp? Is it possible for us to just change the whole thing to use those units?

ctessum commented:
But that can be a separate pull request

ctessum commented:
Here's an additional comment from @jialinl6 

I'm reviewing the FullChem version alongside Xiaokai's SI-unit version to check whether all unit conversions have been applied correctly. One part that seems questionable to me is the handling of the [multiplication of number_density](https://github.com/xk-y/GasChem.jl/blob/857bec3204e1ba4e4d3e45e20e143459b9395790/src/geoschem_ratelaws.jl#L146). For example, in third-body reactions:

Originally, in [geoschem](https://github.com/geoschem/geos-chem/blob/f4c75742e13da440b34c4968797aa150ce114113/KPP/fullchem/gckpp_Rates.F90#L240), the logic behind multiplying by number density is to adjust the units. In a third-body function, the low-pressure rate constant k0_T has units of cm⁶·molec⁻²·s⁻¹. Multiplying it by number density (in molec/cm³) reduces the units to cm³·molec⁻¹·s⁻¹, matching those of the high-pressure rate constant kinf_T. This ensures that the resulting xyrate is unitless. I also found the same issue in SuperFast’s implementation of the third-body reaction function.

One way to fix this is to ensure that all units are consistently in cm³ instead of m³, as in GEOS-Chem—especially since the final output concentrations are in ppb, which is not an SI unit, and not in molec/cm³ or molec/m³. Alternatively, converting everything to SI units (i.e., m³) is also possible, but this might introduce additional, unnecessary bugs.

The unit test bug Xiaokai mentioned: It seems that as long as a non-SI unit is used in an expression, it will give an unmatched unit and value.
But we can keep the non-SI unit by specifying the unit in the substitute function:

jialinl6 commented:
> Here's an additional comment from @jialinl6
> 
> I'm reviewing the FullChem version alongside Xiaokai's SI-unit version to check whether all unit conversions have been applied correctly. One part that seems questionable to me is the handling of the [multiplication of number_density](https://github.com/xk-y/GasChem.jl/blob/857bec3204e1ba4e4d3e45e20e143459b9395790/src/geoschem_ratelaws.jl#L146). For example, in third-body reactions:
> 
> Originally, in [geoschem](https://github.com/geoschem/geos-chem/blob/f4c75742e13da440b34c4968797aa150ce114113/KPP/fullchem/gckpp_Rates.F90#L240), the logic behind multiplying by number density is to adjust the units. In a third-body function, the low-pressure rate constant k0_T has units of cm⁶·molec⁻²·s⁻¹. Multiplying it by number density (in molec/cm³) reduces the units to cm³·molec⁻¹·s⁻¹, matching those of the high-pressure rate constant kinf_T. This ensures that the resulting xyrate is unitless. I also found the same issue in SuperFast’s implementation of the third-body reaction function.
> 
> One way to fix this is to ensure that all units are consistently in cm³ instead of m³, as in GEOS-Chem—especially since the final output concentrations are in ppb, which is not an SI unit, and not in molec/cm³ or molec/m³. Alternatively, converting everything to SI units (i.e., m³) is also possible, but this might introduce additional, unnecessary bugs.
> 
> The unit test bug Xiaokai mentioned: It seems that as long as a non-SI unit is used in an expression, it will give an unmatched unit and value. But we can keep the non-SI unit by specifying the unit in the substitute function:
> 
> (There are also some images that I wasn't able to copy. @jialinl6 would you be willing to add those?)

<img width="624" alt="Screenshot 2025-07-08 at 12 16 00" src="https://github.com/user-attachments/assets/0fe7dc6e-31b1-477e-8ece-2d3ce581c8f1" />
<img width="806" alt="Screenshot 2025-07-08 at 12 16 14" src="https://github.com/user-attachments/assets/c0a4d4c9-f608-4285-b3f7-efd60a2f8f83" />

jialinl6 commented:
> Here's an additional comment from @jialinl6
> 
> I'm reviewing the FullChem version alongside Xiaokai's SI-unit version to check whether all unit conversions have been applied correctly. One part that seems questionable to me is the handling of the [multiplication of number_density](https://github.com/xk-y/GasChem.jl/blob/857bec3204e1ba4e4d3e45e20e143459b9395790/src/geoschem_ratelaws.jl#L146). For example, in third-body reactions:
> 
> Originally, in [geoschem](https://github.com/geoschem/geos-chem/blob/f4c75742e13da440b34c4968797aa150ce114113/KPP/fullchem/gckpp_Rates.F90#L240), the logic behind multiplying by number density is to adjust the units. In a third-body function, the low-pressure rate constant k0_T has units of cm⁶·molec⁻²·s⁻¹. Multiplying it by number density (in molec/cm³) reduces the units to cm³·molec⁻¹·s⁻¹, matching those of the high-pressure rate constant kinf_T. This ensures that the resulting xyrate is unitless. I also found the same issue in SuperFast’s implementation of the third-body reaction function.
> 
> One way to fix this is to ensure that all units are consistently in cm³ instead of m³, as in GEOS-Chem—especially since the final output concentrations are in ppb, which is not an SI unit, and not in molec/cm³ or molec/m³. Alternatively, converting everything to SI units (i.e., m³) is also possible, but this might introduce additional, unnecessary bugs.
> 
> The unit test bug Xiaokai mentioned: It seems that as long as a non-SI unit is used in an expression, it will give an unmatched unit and value. But we can keep the non-SI unit by specifying the unit in the substitute function:
> 
> (There are also some images that I wasn't able to copy. @jialinl6 would you be willing to add those?)

I’d like to retract my previous question regarding the multiplication of number density in units of `molec/cm³` when the Arrhenius reaction rate is in units of `ppb⁻¹·s⁻¹`. It turns out that, regardless of the unit system—whether it's `cm³/molec/s` or `ppb⁻¹·s⁻¹`—we only need to multiply by **the value** of the number density (which corresponds to units of molec/cm³).

The final rate constant `k` [inherits its units directly from](https://github.com/xk-y/GasChem.jl/blob/857bec3204e1ba4e4d3e45e20e143459b9395790/src/geoschem_ratelaws.jl#L155) `rlow`, so there’s no need to convert the number density's units; treating it as a unitless multiplier, as done in GEOS-Chem, is correct. 

However, since the value of number density in `molec/cm³` is not constant at different altitudes in GEOS-Chem, it would be better to compute number density for each box model individually rather than using a [fixed value](https://github.com/xk-y/GasChem.jl/blob/857bec3204e1ba4e4d3e45e20e143459b9395790/src/geoschem_fullchem.jl#L1081). In [GEOS-Chem](https://github.com/geoschem/geos-chem/blob/f4c75742e13da440b34c4968797aa150ce114113/GeosCore/calc_met_mod.F90#L543), where `State_Met%AIRDEN(I,J,L)` represents the dry air density in `kg/m³`, which varies with altitude, temperature and pressure. In SuperFast, air density is calculated in [formula](https://github.com/xk-y/GasChem.jl/blob/857bec3204e1ba4e4d3e45e20e143459b9395790/src/SuperFast.jl#L23): `c = A / (R*T/P) `.

xk-y commented:
I have updated the values. These tests seem to check if we have updated the mechanism so they failed because we updated the geoschem reaction rate constants.

ctessum commented:
OK, we can leave them in for now.
